src/integrations/feed/clients.ts

```diff
@@ -0,0 +1,167 @@
+  }
+
+  getCacheKeyPrefix(feedId?: string): string {
+    return `feeds:${feedId || 'global'}`;
+  }
+
+  getCacheKey(userId?: string, feedId?: string): string {
+    return `${this.getCacheKeyPrefix(feedId)}:${userId || 'anonymous'}`;
+  }
+
+  async shouldServeFromCache(
+    offset: number,
+    key: string,
+    feedId?: string,
+  ): Promise<boolean> {
+    if (offset) {
+      return true;
+    }
+    const updateKey = `${this.getCacheKeyPrefix(feedId)}:update`;
+    const [lastGenerated, lastUpdated] = await this.redisPool.execute(
+      async (client) => {
+        return client.mget(`${key}:time`, updateKey);
+      },
+    );
+    const ttl = 3 * 60 * 1000;
+    return !(
+      !lastGenerated ||
+      (lastUpdated && lastUpdated > lastGenerated) ||
+      new Date().getTime() - new Date(lastGenerated).getTime() > ttl
+    );
+  }
+
+  async fetchFromCache(
+    ctx: Context,
+    feedId: string | undefined,
+    config: FeedConfig,
+  ): Promise<FeedResponse | undefined> {
+    const { offset, page_size: pageSize } = config;
+    const key = this.getCacheKey(config.user_id, feedId);
+    const cachePromise = this.redisPool.execute(async (client) => {
+      return client.get(`${key}:posts`);
+    });
+    if (await this.shouldServeFromCache(offset, key, feedId)) {
+      const cachedFeed = JSON.parse(await cachePromise);
+      if (cachedFeed?.length) {
+        const page: string[] | [string, string | undefined][] =
+          cachedFeed.slice(offset, pageSize + offset);
+        // Support legacy cache that contains only post id
+        if (page.length && typeof page[0] === 'string') {
+          return (page as string[]).map((postId) => [postId, undefined]);
+        }
+        return page as [string, string | undefined][];
+      }
+    }
+    return undefined;
+  }
+
+  async updateCache(
+    feedId: string,
+    config: FeedConfig,
+    items: FeedResponse,
+  ): Promise<void> {
+    const key = this.getCacheKey(config.user_id, feedId);
+    await this.redisPool.execute(async (client) => {
+      const pipeline = client.pipeline();
+      pipeline.set(
+        `${key}:time`,
+        new Date().toISOString(),
+        'EX',
+        ONE_DAY_SECONDS,
+      );
+      pipeline.set(
+        `${key}:posts`,
+        JSON.stringify(items),
+        'EX',
+        ONE_DAY_SECONDS,
+      );
+      return await pipeline.exec();
+    });
+  }
+
+  async fetchFeed(ctx, feedId, config): Promise<FeedResponse> {
+    const { offset, page_size: pageSize } = config;
+    try {
+      const cached = await this.fetchFromCache(ctx, feedId, config);
+      if (cached) {
+        return cached;
+      }
+    } catch (e) {
+      ctx?.log.error({ err: e }, 'failed to get feed from cache');
+    }
+    const cloneConfig = { ...config };
+    delete cloneConfig.offset;
+    const feedRes = await this.client.fetchFeed(ctx, feedId, cloneConfig);
+    if (feedRes.length) {
+      setTimeout(() => this.updateCache(feedId, config, feedRes));
```

Comment:
Why `setTimeout`?

Reply:
I don't want to block returning the feed for storing it in the cache. I think it's higher priority to return asap. Wdyt?
Sidenote: it was part of the previous implementation

Reply:
Get that but if `this.updateCache` is a promise and we don't await it will resolve even without `setTimeout` right?

Reply:
Got it, just saying it is not like `setTimeout` moves it to to the "background" it moves it further down event loop stack.

This should do the same thing in sending a response early and promise will still execute async without blocking the response:
```js
if (feedRes.length) {
  this.updateCache(feedId, config, feedRes);
}

return feedRes.slice(offset, pageSize + offset);
```

So everything is the same I just think `setTimeout` is not needed.

Reply:
Looking into the implementation of `updateCache` I see there's a top level `await` there but if there wasn't the situation would have been different. In this case you're right it's almost the same.
But let's imagine the `await` or the async operation was down the line in this function then using `setTimeout` would make a different than not having await.
The event callback doesn't continue to the next event unless the current event finished processing or is now idle because it waits for an I/O operation.
Meaning, `updateCache` would start its execution immediately until the first encounter of an I/O operation. So if before the first encounter there's heavy processing it will delay sending back the response. It's mostly nuances but sometimes it matters. This is at least according to my understanding

Reply:
When talking in scope of this function having or not having `setTimeout` the `return` line will always execute before anything related to `updateCache`. 

When talking in scope of the whole request processing here there is no way of knowing when `setTimeout` callback will get executed or in what state the queues are since `fetchFeed` is also async and there is more async stuff that executed until actual `req.send` is called internally by fastify.. at least not without performance testing. It can also change while we update the code of course.

If it was like this before let's not change it because it will not hurt response time. It's just a potential head scratcher for anyone looking at the code.